### PR TITLE
[OMPIRBuilder][OpenMP][LLVM] Modify and use ReplaceConstant utility in convertTarget

### PR DESCRIPTION
**Pull request by agozillon**
This PR seeks to expand/replace the Constant -> Instruction conversion that needs to occur inside of the OpenMP Target kernel generation to allow kernel argument replacement of uses within the kernel (cannot replace constant uses within constant expressions with non-constants). It does so by making use of the new-ish utility convertUsersOfConstantsToInstructions which is a much more expansive version of what the smaller "version" of the function I wrote does, effectively expanding uses of the input argument that are constant expressions into instructions so that we can replace with the appropriate kernel argument.

Also alters convertUsersOfConstantsToInstructions to optionally leave dead constants alone is necessary when lowering from MLIR as we cannot be sure we can remove the constants at this stage, even if rewritten to instructions the ModuleTranslation may maintain links to the original constants and utilise them in further lowering steps (as when we're lowering the kernel, the module is still in the process of being lowered). This can result in unusual ICEs later. These dead constants can be tidied up later (and appear to be in subsequent lowering from checking with emit-llvm).

The one possible downside to this replacement is that the constant -> instruction rewriting is no longer constrained to within the kernel, it will expand the available uses of an input argument that is constant and has constant uses in the module. This hasn't lowered the correctness of the examples I have tested with (so far at least, it has only increased correctness e.g. test added in this PR), however, it may impact performance, a possibility in the future may be to optionally constrain rewrites of uses of constants in convertUsersOfConstantsToInstructions to a provided llvm::Function.

### CODE REVIEW

**Comment by agozillon** [comment]
As a slight aside, this PR will have some affect on the following PR: https://github.com/llvm/llvm-project/pull/91829 as the common block PR makes use of the function I am replacing here (with the much better and more advanced/comprehensive utility function that now exists) to achieve it's results. So if/when this lands, I'll replicate these changes to the relevant locations in the other PR which should simplify the changeset as well.

**Comment by agozillon** [comment]
Small ping for a reviewer or two for this PR, it'd be greatly appreciated!

**Comment by nikic** [comment]
> The one possible downside to this replacement is that the constant -> instruction rewriting is no longer constrained to within the kernel, it will expand the available uses of an input argument that is constant and has constant uses in the module. This hasn't lowered the correctness of the examples I have tested with (so far at least, it has only increased correctness e.g. test added in this PR), however, it may impact performance, a possibility in the future may be to optionally constrain rewrites of uses of constants in convertUsersOfConstantsToInstructions to a provided llvm::Function.

I'd feel more comfortable with this if it did not change other functions. It seems like supporting this in convertUsersOfConstantsToInstructions would just be one extra check when creating the InstructionWorklist, so maybe just go ahead and do it now?

**Comment by agozillon** [comment]
> > The one possible downside to this replacement is that the constant -> instruction rewriting is no longer constrained to within the kernel, it will expand the available uses of an input argument that is constant and has constant uses in the module. This hasn't lowered the correctness of the examples I have tested with (so far at least, it has only increased correctness e.g. test added in this PR), however, it may impact performance, a possibility in the future may be to optionally constrain rewrites of uses of constants in convertUsersOfConstantsToInstructions to a provided llvm::Function.
> 
> I'd feel more comfortable with this if it did not change other functions. It seems like supporting this in convertUsersOfConstantsToInstructions would just be one extra check when creating the InstructionWorklist, so maybe just go ahead and do it now?

Can do so! I was a bit hesitant to do it without some input as I wasn't sure exactly the best place to put the check to constrain it. But it should only take one extra check and an extra argument for the current function/kernel the conversion should be restricted to (at least from my simplistic understanding of the function)

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 1]
```diff
@@ -5191,17 +5171,23 @@ static Function *createOutlinedFunction(
...
+    // allows us to be cautious as it is perhaps possible the old expression
+    // was used inside of the function but exists and is used externally
+    // (unlikely by the nature of a Constant, but still).
+    // NOTE: We cannot remove dead constants that have been rewritten to
+    // instructions at this stage, we run the risk of breaking later lowering
+    // by doing so as we could still be in the process of lowering the module
+    // from MLIR to LLVM-IR and the MLIR lowering may still require the original
+    // constants we have created rewritten versions of.
+    if (auto *Const = dyn_cast<Constant>(Input))
+      convertUsersOfConstantsToInstructions({Const}, Func, false);
```
Comment (nikic) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5190]:
```suggestion
      convertUsersOfConstantsToInstructions(Const, Func, false);
```
ArrayRef has a single-value overload.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp (hunk 1 above)
Comment (agozillon) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5190]: Ah, thank you, I wasn't aware! So used to brace initializing most things. I'll make this alteration and update in a little bit :-)

**Comment by agozillon** [comment]
> LGTM

Thank you very much! I'll pester one other reviewer for a review/sign-off and then hopefully it'll be good to land.

File: llvm/include/llvm/IR/ReplaceConstant.h [hunk 2]
```diff
@@ -18,10 +18,13 @@ namespace llvm {
 
 template <typename T> class ArrayRef;
 class Constant;
+class Function;
 
 /// Replace constant expressions users of the given constants with
 /// instructions. Return whether anything was changed.
-bool convertUsersOfConstantsToInstructions(ArrayRef<Constant *> Consts);
+bool convertUsersOfConstantsToInstructions(ArrayRef<Constant *> Consts,
```
Comment (jsjodin) [inline llvm/include/llvm/IR/ReplaceConstant.h:25]: Please add descriptions of the new parameters to the function in the comment above.

**Review by jsjodin — APPROVED** [review]
Just a nit, otherwise it looks good to me.

**Comment by agozillon** [comment]
Updated the comment in the last commit to provide details on the arguments! Thank you very much @jsjodin and @nikic I will merge the changes now.

**Pull request merged**